src/slurm_plugin/clustermgtd.py

```diff
@@ -879,46 +906,227 @@ def _terminate_orphaned_instances(self, cluster_instances, ips_used_by_slurm):
+            else:
+                self.num_power_down_nodes += 1
+
+    def get_values(self):
+        # TODO: let's use yield when we have multiple lists and have to iterate on them
+        if self._has_received:
+            self._update_power_up_down_nodes()
+            return [
+                {"name": "slurm_number_power_up_nodes", "value": self.num_power_up_nodes},
+                {"name": "slurm_number_power_down_nodes", "value": self.num_power_down_nodes},
+                {"name": "slurm_number_inactive_nodes", "value": len(self.slurm_inactive_nodes)},
+                {"name": "number_ec2_nodes", "value": len(self.ec2_nodes)},
+            ]
+        return []
+
+
+def _push_metrics_to_cloudwatch(cluster_metrics, exceptions_list):
+
+    try:
+        # FIXME 'StorageResolution' should not be specified I think, here just for faster tests
+        config = cluster_metrics.config
+        current_thread = currentThread()
+        while getattr(current_thread, "keep_running", True):
+            start_time = datetime.now(tz=timezone.utc)
+
+            # get most recent values ([] if no values were uploaded at all (should happen only the first time)
+            metrics = cluster_metrics.get_values()
+
+            # upload metrics
+            # TODO: Change Namespace to Cluster Metrics
+            if metrics:
+                config = cluster_metrics.config
+                cw_client = boto3.client("cloudwatch", region_name=config.region, config=config.boto3_config)
+                cw_client.put_metric_data(
+                    Namespace="Custom Namespace",
+                    MetricData=[
+                        {
+                            "MetricName": metric["name"],
+                            "Dimensions": [{"Name": "ClusterName", "Value": config.cluster_name}],
+                            "Timestamp": start_time,
+                            "Value": metric["value"],
+                            "Unit": "Count",
+                            "StorageResolution": 1,
+                        }
+                        for metric in metrics
+                    ],
+                )
+            else:
+                log.info("No metrics to upload")
+
+            # FIXME What happens if the put_metric_data call takes too much time (more than config.loop_time) ?
+            # At the moment, the delay would simply add up every time... Should we have this or should we
+            # upload faster if we are late? This could be an idea but a problem is that if it happens because of
+            # performance problems (instance very busy?), then, it would make it even more difficult for the instance
+            # to do its work.
+            sleep_remaining_loop_time(config.loop_time, start_time)
+    except Exception as e:
+        exceptions_list.append(ClusterMetricsException(e))
+        return
+    log.info("Leaving _push_metrics_to_cloudwatch")
+
+
+class RestartableThread(Thread):
```

Comment:
have you considered configuring these threads with the `daemon=true` flag so that if the main program exits it does not hang but simply exits all daemon threads?

Reply:
Using this now, thanks for the suggestion!